Rbac (#140)
percona/everest-doc: docs/administer/api_rbac.md

> # Navigating the API changes for RBAC

Reviewer: I think its important to mention in the heading that these are **breaking** API changes. Maybe something like:
- Navigating the breaking API changes for RBAC
- Breaking API changes in Everest 1.2.0

Author: Sure

Author: Done

> Starting with Percona Everest v1.2.0, breaking changes are being made to the API for monitoring instances and backup storage resources. These changes are:

Reviewer: Please consider adding backticks for the resource names, like so:

```suggestion
Starting with Percona Everest v1.2.0, breaking changes are being made to the API for `monitoring-instances` and `backup-storage` resources. These changes are:
```

- Before the release of Percona Everest 1.2.0, these resources were globally scoped, but now they will be specific to namespaces. 

- The database clusters will only be able to use monitoring instances and backup storages that are located within the same namespace as the cluster. The system used a `.spec.allowedNamespaces` field to control access to these global resources. This field determined the namespaces where the resource could be accessed, providing a certain degree of access control.

- With the update to Percona Everest v1.2.0, the shift from global scoping to namespacing these resources marks a significant change in how access control is managed, enhancing security by ensuring that resources are only accessible within their designated namespaces.

> ## Challenges of globally spaced namespaces

Reviewer: TBH this heading doesn't make much sense to me. 😅 

Please consider something like:
- Challenges with globally scoped resources
- Why?

Author: Fixed.

Reviewer: I still see the older title `Challenges of globally spaced namespaces`

Author: Fixed

> The use of the `allowedNamespaces` field for globally scoped resources presented challenges when integrating with the core RBAC model. In order to resolve this inconsistency and align with the RBAC framework, backup storage and monitoring instances are now namespaced resources. This change ensures that all resources adhere to the same RBAC model, resulting in a more consistent and manageable access control structure across the entire system.

Reviewer: I think we should start by mentioning that 1.2.0 introduces RBAC, and then describe how the current model poses challenges when integrating with it

Reviewer: I feel that this is a heavily trimmed down version of the ["Why?"](https://www.notion.so/percona/Breaking-API-changes-in-Everest-v1-2-0-619c0a3540bb4038bb84e2d493b61646?pvs=4#303826c1e4084f7798f636e4d7263bc7) section from the Notion doc. It would be great if you'd consider taking a look again and see if we can paste it here as-is (or with any changes to the wording you feel is necessary..).

IMO this information is important because as a developer, I wouldn't be expecting a breaking API change for a minor upgrade, so it's important we provide adequate justification 😅

Author: Fixed

##  Changes in the Percona Everest APIs

The APIs have been updated with the following modifications:

- The existing APIs for backup storage and monitoring instances are deprecated. Now, you should use the API path prefixed with `/namespaces/{namespace}`.

    ??? example "Example"

        ```/v1/backup-storages``` is now

        ```/v1/namespaces/{namespace}/backup-storages```

        Check out the [API](https://percona-everest.readme.io/reference/getkubernetesclusterresources) documentation for more details.

- The `.spec.allowedNamespaces` field has been deprecated. Access control for these resources is now managed through the RBAC policy.

### Migrating to Percona Everest 1.2.0

> After the upgrade to Percona Everest 1.2.0, you will only be able to access these resources through the new API endpoints.
> 
> Need more details? Check out the [upgrade](../upgrade_with_cli.md#upgrading-to-percona-everest-120) section.

Reviewer: It is important to mention explicitly that all their existing `monitoring-configs` and `backup-storages` will be automatically migrated.

```suggestion
When upgrading to 1.2.0, all your existing backup-storages and monitoring-instances will be automatically migrated to the namespaces specified in their .spec.allowedNamespaces fields. After the upgrade, these resources will be accessible exclusively through the new API endpoints.

Need more details? Check out the [upgrade](../upgrade_with_cli.md#upgrading-to-percona-everest-120) section.
```

# Navigating the breaking API changes for RBAC

Starting with Percona Everest v1.2.0, breaking changes are being made to the API for `monitoring-instances` and `backup-storage` resources. These changes are:

- Before the release of Percona Everest 1.2.0, these resources were globally scoped, but now they will be specific to namespaces. 

> - The database clusters can only use monitoring instances and backup storages located within the same namespace as the cluster. The system used a `.spec.allowedNamespaces` field to control access to these global resources. This field determined the namespaces where the resource could be accessed, providing a certain degree of access control.

Reviewer: Please add backticks where necessary so that its consistent throughout the document

```suggestion
- The database clusters can only use `monitoring-instances` and `backup-storages` located within the same namespace as the cluster. The system used a `.spec.allowedNamespaces` field to control access to these global resources. This field determined the namespaces where the resource could be accessed, providing a certain degree of access control.
```

- With the update to Percona Everest v1.2.0, the shift from global scoping to namespacing these resources marks a significant change in how access control is managed, enhancing security by ensuring that resources are only accessible within their designated namespaces.

> ## Challenges with globally spaced namespaces

Reviewer: TBH this heading doesn't make much sense to me. 😅

Please consider something like:

Challenges with globally scoped resources
Why?

In Percona Everest v1.2.0, we’ve rolled out Role-Based Access Control (RBAC) to enhance security and provide more granular control over the access privileges for specific resources within the system. This implementation provides fine-grained control over which users and user groups can access particular resources within the system. 

The RBAC model functions on the principle that all resources are organized into namespaces. This enables a well-structured and hierarchical arrangement of resources, simplifying the management of access rights according to the namespace to which a resource is associated.

> Prior to Percona Everest version 1.2.0, certain resources such as `backup-storages` and `monitoring-instances` were not organized into namespaces but were accessible globally. To enforce access restrictions on these globally scoped resources, the system utilized a .spec.allowedNamespaces field. The .spec.allowedNamespaces field specifies the namespaces within which the resource can be accessed, giving you certain level of control.

Reviewer: ```suggestion
Prior to Percona Everest version 1.2.0, certain resources such as `backup-storages` and `monitoring-instances` were not organized into namespaces but were accessible globally. To enforce access restrictions on these globally scoped resources, the system utilized a `.spec.allowedNamespaces` field. This field specifies the namespaces within which the resource can be accessed, giving you certain level of control.
```

> Using the `allowedNamespaces` field for globally scoped resources presented challenges when integrating with the core RBAC model. To fix this and align with the RBAC framework, `backup-storages`, and `monitoring-instances` are now namespaced resources. This ensures that all resources conform to the same RBAC model, which results in a consistent and manageable access control structure across the system.

Reviewer: ```suggestion
Using the `.spec.allowedNamespaces` field for globally scoped resources presented challenges when integrating with the core RBAC model. To fix this and align with the RBAC framework, `backup-storages` and `monitoring-instances` are now namespaced resources. This ensures that all resources conform to the same RBAC model, which results in a consistent and manageable access control structure across the system.
```

##  Changes in the Percona Everest APIs

The APIs have been updated with the following modifications:

- The existing APIs for backup storage and monitoring instances are deprecated. Now, you should use the API path prefixed with `/namespaces/{namespace}`.

    ??? example "Example"

        ```/v1/backup-storages``` is now

        ```/v1/namespaces/{namespace}/backup-storages```

        Check out the [API](https://percona-everest.readme.io/reference/getkubernetesclusterresources) documentation for more details.

- The `.spec.allowedNamespaces` field has been deprecated. Access control for these resources is now managed through the RBAC policy.

- `database-clusters` can now only reference `backup-storages` and `monitoring-instances` that are created within the same namespace as the `database-cluster`.

### Migrating to Percona Everest 1.2.0

> When upgrading to 1.2.0, all your existing backup-storages and monitoring-instances will be automatically migrated to the namespaces specified in their .spec.allowedNamespaces fields. After the upgrade, these resources will be accessible exclusively through the new API endpoints.

Reviewer: ```suggestion
When upgrading to 1.2.0, all your existing `backup-storages` and `monitoring-instances` will be automatically migrated to the namespaces specified in their respective `.spec.allowedNamespaces` fields. After the upgrade, these resources will be accessible exclusively through the new API endpoints.
```
